File: Sources/Stockflow.Webserver/Scenarios/Scenario.cs
namespace Stockflow.Webserver.Scenarios;

/// <summary>
/// Schema scenario file-based, mirror del formato §6.1 di ARCHITECTURE_StockFlow_v0.3.
/// I campi non ancora supportati dalla simulazione (Budget, SkuCatalog, OrderProfile,
/// Objectives, AvailableComponents, Duration) sono pass-through: persistiti ma
/// ignorati dal loader con warning a log.
/// </summary>
public sealed record Scenario
{
    public required string Id          { get; init; }
    public required string Name        { get; init; }
    public          string? Description { get; init; }

    public GridSize?           GridSize            { get; init; }
    public int?                Budget              { get; init; }
    public IReadOnlyList<string>? AvailableComponents { get; init; }
    public IReadOnlyList<PreplacedComponent>? PreplacedComponents { get; init; }
    public SkuCatalog?         SkuCatalog          { get; init; }
    public OrderProfile?       OrderProfile        { get; init; }
    public IReadOnlyDictionary<string, Objective>? Objectives { get; init; }
    public int?                Duration            { get; init; }
}

public sealed record GridSize(int Width, int Height);

public sealed record PreplacedComponent(string Type, int[] Position, string Direction);

public sealed record SkuCatalog(int Count, IReadOnlyDictionary<string, SkuClass> Classes);

public sealed record SkuClass(int Percentage, string AccessFrequency);

public sealed record OrderProfile(
    int            BaseRate,
    int            PeakRate,
    int            PeakStartTime,
    int            PeakDuration,
    LinesPerOrder? LinesPerOrder,
    int            Deadline);

public sealed record LinesPerOrder(int Min, int Max);

public sealed record Objective(string Type, double? Value = null);

File: Sources/Stockflow.Console/src/app/core/models/scenario.ts
/**
 * Mirror dei record C# Stockflow.Webserver.Scenarios.* (formato §6.1
 * di ARCHITECTURE_StockFlow_v0.3). I campi non ancora supportati dalla
 * simulazione sono opzionali pass-through.
 */

export interface ScenarioSummary {
  id: string;
  name: string;
  description?: string | null;
}

export interface Scenario {
  id: string;
  name: string;
  description?: string | null;
  gridSize?: GridSize | null;
  budget?: number | null;
  availableComponents?: string[] | null;
  preplacedComponents?: PreplacedComponent[] | null;
  skuCatalog?: SkuCatalog | null;
  orderProfile?: OrderProfile | null;
  objectives?: Record<string, Objective> | null;
  duration?: number | null;
}

export interface GridSize {
  width: number;
  height: number;
}

export interface PreplacedComponent {
  type: string;
  position: number[]; // [x, y]
  direction: string;
}

export interface SkuCatalog {
  count: number;
  classes: Record<string, SkuClass>;
}

export interface SkuClass {
  percentage: number;
  accessFrequency: string;
}

export interface OrderProfile {
  baseRate: number;
  peakRate: number;
  peakStartTime: number;
  peakDuration: number;
  linesPerOrder?: LinesPerOrder | null;
  deadline: number;
}

export interface LinesPerOrder {
  min: number;
  max: number;
}

export interface Objective {
  type: string;
  value?: number | null;
}
